File: src/deus_profiling.c
/**
* @file deus_profiling.c
* @brief Calculation of the standard deviation of numbers from the inputfile.
*
* @author Pavol Szepsi
*
* @version 1.5.0
* @date 14.4.2019
*/

# include <stdio.h>
# include <math.h>

/**
* @brief Function calculates standard deviation.
*
* @param file Input file with numbers.
*
* @return Standard deviation of numbers from input.
*/
double StandardDeviation(char *file) {
	FILE *pFile;
	pFile = fopen(file, "r");
	if (pFile == NULL)
		return 1;

	double x, avgX;
	double sumX = 0;
	double sumX2 = 0;
	int N = 0;

	while (fscanf(pFile, "%lf", &x) != EOF) {
		sumX += x;
		sumX2 += pow(x, 2);
		N += 1;
	}
	fclose(pFile);

	avgX = sumX / N;
	return sqrt(1.0 / (N - 1) * (sumX2 - (N * pow(avgX, 2))));
}

/**
* @brief Main function calling.
*
* @param argc Number of arguments.
* @param argv[] Array of arguments.
*
* @return 0 - Program finished successfully.
* @return 1 - error - File couldn't be opened!
* @return 2 - error - Need filename as the first argument!
* @return 3 - error - Too many arguments!
*/
int main(int argc, char *argv[]) {
	if (argc == 1) {
		fprintf(stderr, "Need filename as the first argument!\n");
		return 2;
	}

	if (argc > 2) {
		fprintf(stderr, "Too many arguments!\n");
		return 3;
	}

	if (StandardDeviation(argv[1]) == 1) {
        fprintf(stderr, "File couldn't be opened!\n");
        return 1;
	}

    printf("%lf\n", StandardDeviation(argv[1]));
	return 0;
}

/**
* @}
*/

File: src/deus_profiling.py
#!/usr/bin/python
# -*- coding: utf-8 -*-


###############################################################################
# @file deus_profiling.py
# @brief Calculation of the standard deviation of numbers from the inputfile.
#
# @author Pavol Szepsi (xszeps00)
#
# @version 2.0.0
# @date 21.4.2019
###############################################################################


import sys
import deus_math_lib


##
# @brief Function calculates standard deviation.
#
# @return Standard deviation of numbers from input
def StandardDeviation():
    sumX = 0
    sumX2 = 0
    N = 0

    for x in sys.stdin:
        x = float(x)
        sumX = deus_math_lib.deus_sum(sumX, x)
        sumX2 = deus_math_lib.deus_sum(sumX2, deus_math_lib.deus_pow(x, 2))
        N = deus_math_lib.deus_sum(N, 1)

    avgX = deus_math_lib.deus_div(sumX, N)
    res = deus_math_lib.deus_root(deus_math_lib.deus_mult(deus_math_lib.deus_div(1, deus_math_lib.deus_sub(N, 1)), deus_math_lib.deus_sub(sumX2, deus_math_lib.deus_mult(N, deus_math_lib.deus_pow(avgX, 2)))), 2)
    return res


x = StandardDeviation()
print(x)


###############################################################################
# End of the deus_profiling.py
###############################################################################
